OpenSim/Simulation/Wrap/PathWrapPoint.cpp

```diff
@@ -0,0 +1,124 @@
+/* -------------------------------------------------------------------------- *
+ *                           OpenSim:  PathWrapPoint.cpp                      *
+ * -------------------------------------------------------------------------- *
+ * The OpenSim API is a toolkit for musculoskeletal modeling and simulation.  *
+ * See http://opensim.stanford.edu and the NOTICE file for more information.  *
+ * OpenSim is developed at Stanford University and supported by the US        *
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2017 Stanford University and the Authors                *
+ * Author(s): Peter Loan                                                      *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "PathWrapPoint.h"
+
+OpenSim::PathWrapPoint::PathWrapPoint() = default;
+
+void OpenSim::PathWrapPoint::extendAddToSystem(SimTK::MultibodySystem& system) const
+{
+    Super::extendAddToSystem(system);
+
+    _wrapPath = addCacheVariable("wrap_path", Array<SimTK::Vec3>{}, SimTK::Stage::Position);
+    _wrapPathLength = addCacheVariable("wrap_path_length", 0.0, SimTK::Stage::Position);
+    _location = addCacheVariable("wrap_location", SimTK::Vec3{}, SimTK::Stage::Position);
+}
+
+const OpenSim::Array<SimTK::Vec3>& OpenSim::PathWrapPoint::getWrapPath(const SimTK::State& s) const
+{
+    return getCacheVariableValue(s, _wrapPath);
+}
+
+void OpenSim::PathWrapPoint::setWrapPath(const SimTK::State& s, const Array<SimTK::Vec3>& src) const
+{
+    updCacheVariableValue(s, _wrapPath) = src;
```

Comment:
Patterns like this, and in other OpenSim components, need to be fixed by adding high-order functions like:

- `ApplyCacheVariableChange`
- `FetchCacheVariableValue(cv, updateFunction);`

Where there's only one cache variable lookup. The lookup is cheaper now, due to previous cache variable changes, but it still irks me that there are two lookups happening when one should do.